### PR TITLE
Customizable signup fields

### PR DESCRIPTION
**Pull request by haverchuck**
*Issue #, if available:*
Beta test for:
#1911 
#1901 
#935 
#1871 

*Description of changes:*
Adds signup customization for Angular, React, React Native and Ionic.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/aws-amplify-react/src/Auth/index.jsx [hunk 1]
```diff
@@ -34,7 +34,7 @@ export * from './Provider';
 
 import Greetings from './Greetings';
 
-export function withAuthenticator(Comp, includeGreetings = false, authenticatorComponents = [], federated = null, theme = null) {
+export function withAuthenticator(Comp, includeGreetings = false, authenticatorComponents = [], federated = null, theme = null, signUpConfig = {}) {
```
Comment (powerful23) [inline packages/aws-amplify-react/src/Auth/index.jsx]: Can we refactor this hoc function to something like:
```js
withAuthenticator(Comp, {
    includingGreetings: false,
    // ...
});
```

File: packages/aws-amplify-react/src/Auth/SignUp.jsx [hunk 2]
```diff
@@ -33,103 +33,123 @@ import {
...
+        });
 
-        Auth.signUp(signup_info).then(() => this.changeState('confirmSignUp', username))
+        Auth.signUp(signup_info).then((data) => {
+            this.changeState('confirmSignUp', data.user.username)
+        })
         .catch(err => this.error(err));
     }
 
     showComponent(theme) {
```
Comment (powerful23) [inline packages/aws-amplify-react/src/Auth/SignUp.jsx]: Can you move this function to the bottom?

File: packages/aws-amplify-react-native/src/Auth/SignUp.js [hunk 3]
```diff
@@ -33,71 +34,90 @@ import {
...
+
+        console.log('signup_info', signup_info)
+
+        Auth.signUp(signup_info).then((data) => {
+            this.changeState('confirmSignUp', data.user.username)
+        })
+        .catch(err => this.error(err));
     }
 
     showComponent(theme) {
```
Comment (powerful23) [inline packages/aws-amplify-react-native/src/Auth/SignUp.js]: Can we move this function to the bottom?

File: packages/aws-amplify-react-native/src/Auth/SignUp.js [hunk 4]
```diff
@@ -118,4 +138,95 @@ export default class SignUp extends AuthPiece {
...
+          this.signUpFields = this.defaultSignUpFields;
+        }
+    }
+
+    needPrefix(key) {
+        const field = this.signUpFields.find(e => e.key === key);
+        if (key.indexOf('custom:') !== 0) {
+          return field.custom ;
+        } else if (key.indexOf('custom:') === 0 && field.custom === false) {
+          this.amplifyService.logger('SignUpComponent', 'WARN')
```
Comment (powerful23) [inline packages/aws-amplify-react-native/src/Auth/SignUp.js]: What is this `amplifyService`?

File: packages/aws-amplify-react/src/Auth/SignUp.jsx [hunk 5]
```diff
@@ -147,4 +167,103 @@ export default class SignUp extends AuthPiece {
...
+          this.signUpFields = this.defaultSignUpFields;
+        }
+    }
+
+    needPrefix(key) {
+        const field = this.signUpFields.find(e => e.key === key);
+        if (key.indexOf('custom:') !== 0) {
+          return field.custom ;
+        } else if (key.indexOf('custom:') === 0 && field.custom === false) {
+          this.amplifyService.logger('SignUpComponent', 'WARN')
```
Comment (powerful23) [inline packages/aws-amplify-react/src/Auth/SignUp.jsx]: What is this `amplifyService`?

File: packages/aws-amplify-react/src/Auth/index.jsx (hunk 1 above)
Comment (haverchuck) [reply packages/aws-amplify-react/src/Auth/index.jsx]: Yes I was thinking that, but I dont' want to do it at this time. We're trying to get the updated components into beta this week.

File: packages/aws-amplify-react/src/Auth/SignUp.jsx (hunk 2 above)
Comment (haverchuck) [reply packages/aws-amplify-react/src/Auth/SignUp.jsx]: Done.

File: packages/aws-amplify-react-native/src/Auth/SignUp.js (hunk 3 above)
Comment (haverchuck) [reply packages/aws-amplify-react-native/src/Auth/SignUp.js]: Done.

File: packages/aws-amplify-react-native/src/Auth/SignUp.js (hunk 4 above)
Comment (haverchuck) [reply packages/aws-amplify-react-native/src/Auth/SignUp.js]: Copy & paste error from Angular... fixed now.

File: packages/aws-amplify-react/src/Auth/SignUp.jsx (hunk 5 above)
Comment (haverchuck) [reply packages/aws-amplify-react/src/Auth/SignUp.jsx]: Copy & paste error from Angular... fixed now.

**Pull request merged**